### PR TITLE
Drivetrain hw for ros2 galactic

### PR DESCRIPTION
**Pull request by wmmc88**
drivetrain URDF can now be displayed in rviz using: `ros2 launch uwrt_mars_rover_drivetrain_description display.launch.py`

You can test the roscontrol code locally using `ros2 launch uwrt_mars_rover_drivetrain_hw drivetrain.launch.py`. You'll need something to publish stamped twists as a data source for the drivetrain controller to actually see it print the correct values in read/write. You'll want to launch the node using debug logger level to actually see the read/write prints

chained to #156 

many fixes to CI problems that were exposed while trying to get this PR to pass CI

Followups(in this order):
Transmission Support
CAN for real hw (or the udp stuff for the prototype)
gazebo support

### CODE REVIEW

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:61]: these values should be tuned while testing on the actual rover

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (niiquaye) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:21]: aren't you checking State interfaces here not command interfaces ..... they are the number of state and command so not an issue ... but just curious

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (niiquaye) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:29]: lol I was thinking same .... I just defined this in my class
`static constexpr char HW_IF_IQ_CURRENT[] = "iq_current;"` .... just copy and paste from what ros2 does

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/urdf/drivetrain.macro.xacro
Comment (niiquaye) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/urdf/drivetrain.macro.xacro:89]: man this xacro makes it look so immaculate and beautiful

Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/urdf/drivetrain.macro.xacro:89]: yup! thats the beauty of xacro. suuuuuuuper clean

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:29]: yeah. I think its best to keep it as just a string until we create a package for it to avoid duplicate definitions everywhere. Probably like a top level `uwrt_mars_rover_hw` package like what i did for shared custom interfaces in the `ros1` branch

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:15]: that sounds overly complex for what it accomplishes(no need for function pointers if ur theyre never gonna change). YAGNI. 

If we want joint specific loggers, I'll just hold a `logger_` instance in the class and initialize it to the joint name in the `on_init`.

Comment (niiquaye) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:15]: hmm that makes a lot of sense with your approach

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml:14]: TODO: all these variables need to be updated to values provided by mech

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:30]: this has been merged already. just waiting on binary release

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:34]: TODO: add pr reference

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py:30]: Create issue for this

File: .github/workflows/ci.yaml
Comment (wmmc88) [inline .github/workflows/ci.yaml:16]: fyi @niiquaye: 

We need to add every new package we want tested into this list. I added a thing into the pr template to address this

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py
Comment (niiquaye) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py:30]: custom controller manager .... with real time capabilities potentially?

File: .github/workflows/ci.yaml
Comment (niiquaye) [reply .github/workflows/ci.yaml]: alright sounds good

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py:30]: no not a custom controller manager. just instead of using the node provided by `ros2_control`, we use our own node that is almost the same, except makes the appropriate syscalls to get rt performance on a linux machine with the prempt rt patch. see this:
https://github.com/uwrobotics/uwrt_mars_rover/blob/ros1/uwrt_mars_rover_control/src/uwrt_mars_rover_control_node.cpp

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/CMakeLists.txt
Comment (wmmc88) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/CMakeLists.txt:1]: @niiquaye we should probably use this file as like a template for other Cmakelists in the repo as it has all the necessary things to build, run tests, etc. if we keep all of the cmake stuff looking somewhat the same, itll be eaasier to maintain

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml:14]: image

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/CMakeLists.txt
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/CMakeLists.txt:1]: i updated the template in our README. all cmakelists should follow this imo

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp]: implemented.

**Comment by wmmc88** [comment]
Green CI means no bugs, right? /s

Ready for review.

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:61]: https://github.com/uwrobotics/uwrt_mars_rover/issues/177

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py:39]: #178

**Review by niiquaye — APPROVED** [review]
What a beauty - Nice Job!

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:30]: Version 1.3.0 has released and has been synced upstream:
https://github.com/ros-controls/ros2_controllers/blob/master/diff_drive_controller/CHANGELOG.rst
https://index.ros.org/p/diff_drive_controller/github-ros-controls-ros2_controllers/#galactic

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/src/uwrt_mars_rover_drivetrain_hw_actuator_interface.cpp:32]: #179

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/config/physical_parameters.yaml]: updated

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml
Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/config/drivetrain_controllers.yaml:34]: added ref to discussion

**Comment by wmmc88** [comment]
Ok. Did one last round of cleanup (and testing) and opened relevant followup issues. Should be good to merge now, unless anyone else has comments

File: uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/include/uwrt_mars_rover_drivetrain_hw/visibility_control.hpp
Comment (niiquaye) [inline uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/include/uwrt_mars_rover_drivetrain_hw/visibility_control.hpp:1]: Not important but `#pragma once`

Comment (wmmc88) [reply uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/include/uwrt_mars_rover_drivetrain_hw/visibility_control.hpp]: fixed!

**Review by niiquaye — APPROVED** [review]

**Pull request merged**
